hack/generator/pkg/astmodel/conversion_function_builder.go

```diff
@@ -0,0 +1,466 @@
+
+func (params ConversionParameters) withAdditionalConversionContext(t Type) ConversionParameters {
+	result := params.copy()
+	result.ConversionContext = append(result.ConversionContext, t)
+
+	return result
+}
+
+func (params ConversionParameters) WithSource(source dst.Expr) ConversionParameters {
+	result := params.copy()
+	result.Source = source
+
+	return result
+}
+
+func (params ConversionParameters) WithSourceType(t Type) ConversionParameters {
+	result := params.copy()
+	result.SourceType = t
+
+	return result
+}
+
+func (params ConversionParameters) WithDestination(destination dst.Expr) ConversionParameters {
+	result := params.copy()
+	result.Destination = destination
+
+	return result
+}
+
+func (params ConversionParameters) WithDestinationType(t Type) ConversionParameters {
+	result := params.copy()
+	result.DestinationType = t
+
+	return result
+}
+
+func (params ConversionParameters) WithAssignmentHandler(
+	assignmentHandler func(result dst.Expr, destination dst.Expr) dst.Stmt) ConversionParameters {
+	result := params.copy()
+	result.AssignmentHandler = assignmentHandler
+
+	return result
+}
+
+func (params ConversionParameters) AssignmentHandlerOrDefault() func(destination, source dst.Expr) dst.Stmt {
+	if params.AssignmentHandler == nil {
+		return AssignmentHandlerAssign
+	}
+	return params.AssignmentHandler
+}
+
+// TODO: Consider a better way to do this, using KnownLocalsSet?
+// CountArraysAndMapsInConversionContext returns the number of arrays/maps which are in the conversion context.
+// This is to aid in situations where there are deeply nested conversions (i.e. array of map of maps). In these contexts,
+// just using a simple assignment such as "elem := ..." isn't sufficient because elem my already have been defined above by
+// an enclosing map/array conversion context. We use the depth to do "elem1 := ..." or "elem7 := ...".
+func (params ConversionParameters) CountArraysAndMapsInConversionContext() int {
+	result := 0
+	for _, t := range params.ConversionContext {
+		switch t.(type) {
+		case *MapType:
+			result += 1
+		case *ArrayType:
+			result += 1
+		}
+	}
+
+	return result
+}
+
+type ConversionHandler func(builder *ConversionFunctionBuilder, params ConversionParameters) []dst.Stmt
+
+// TODO: There feels like overlap between this and the Storage Conversion Factories? Need further thinking to combine them?
+// ConversionFunctionBuilder is used to build a function converting between two similar types.
+// It has a set of built-in conversions and can be configured with additional conversions.
+type ConversionFunctionBuilder struct {
+	// TODO: Better way to let you fuss with this? How can you pick out what I've already put in here to overwrite it?
+	conversions []ConversionHandler
+
+	IdFactory             IdentifierFactory
+	CodeGenerationContext *CodeGenerationContext
+}
+
+func NewConversionFunctionBuilder(idFactory IdentifierFactory, codeGenerationContext *CodeGenerationContext) *ConversionFunctionBuilder {
```

Comment:
It'd be nice to both: "have an order" to things that are processed, but also allow them to be overwritable/movable (like `TypeVisitor`'s are).

I didn't spend a ton of time thinking how to accomplish that for this first pass but possibly something to think about if this is going to be used more broadly.